File: coco_scala/src/main/scala/exp/m4/M4.scala
package exp.m4

trait Exp[T] extends exp.m2.Exp[T] with Factory[T] {
  def simplify : exp.Exp[T]
  def truncate(level:Int) : Unit
  def collect : List[Double]
}

trait Factory[T] extends exp.m3.Factory[T] {
  implicit override def convert(toConvert:exp.Exp[T]) : exp.m4.Exp[T]
}

trait Lit[T] extends exp.m2.Lit[T] with Exp[T] {
  def simplify : exp.Exp[T] = this
  def truncate(level:Int) = ()
  def collect : List[Double] = List(this.value)
}

trait BinaryExp[T] extends Exp[T] {
  var _left:exp.Exp[T]
  var _right:exp.Exp[T]
  def left = _left
  def right = _right

  def truncate(level:Int) = {
    if (level > 0) {
      left.truncate(level-1)
      right.truncate(level-1)
    } else {
      _left = lit(left.eval)
      _right = lit(right.eval)
    }
  }
}

trait Add[T] extends exp.m2.Add[T] with BinaryExp[T] {
  def simplify : exp.Exp[T] = {
    if (left.eval + right.eval == 0) {
      this.lit(0)
    } else if (left.eval == 0) {
      right.simplify
    } else if (right.eval == 0) {
      left.simplify
    } else {
      this.add(left.simplify, right.simplify)
    }
  }

  def collect : List[Double] = left.collect ++ right.collect
}

trait Sub[T] extends exp.m2.Sub[T] with BinaryExp[T] {
  def simplify : exp.Exp[T] = {
    if (left.eval - right.eval == 0) {
      this.lit(0)
    } else if (right.eval == 0) {
      left.simplify
    } else if (left.eval == 0) {
      this.neg(right.simplify)     // can use neg from earlier
    } else {
      this.sub(left.simplify, right.simplify)
    }
  }

  def collect : List[Double] = left.collect ++ right.collect
}

trait Mult[T] extends exp.m3.Mult[T] with BinaryExp[T] {
  def simplify : exp.Exp[T] = {
    if (left.eval == 0 || right.eval == 0) {
      this.lit(0)
    } else if (right.eval == 1) {
      left.simplify
    } else if (left.eval == 1) {
      right.simplify     // can use neg from earlier
    } else {
      this.mult(left.simplify, right.simplify)
    }
  }

  def collect : List[Double] = left.collect ++ right.collect
}

trait Neg[T] extends exp.m3.Neg[T] with Exp[T] {
  var _inner:exp.Exp[T]
  def inner:exp.Exp[T] = _inner

  def truncate(level:Int) = {
    if (level > 0) {
      inner.truncate(level-1)
    } else {
      _inner = lit(inner.eval)
    }
  }

  def simplify : exp.Exp[T] = {
    if (inner.eval == 0 ) {
      this.lit(0)
    } else {
      this.neg(inner.simplify)
    }
  }

  def collect : List[Double] = inner.collect
}

trait Divd[T] extends exp.m3.Divd[T] with BinaryExp[T] {
  def simplify : exp.Exp[T] = {
    if (left.eval == 0 ) {
      this.lit(0)
    } else if (right.eval == 1) {
      left.simplify
    } else if (left.eval == right.eval) {
      this.lit(1)
    } else if (left.eval == -right.eval) {
      this.lit(-1)
    } else {
      this.divd(left.simplify, right.simplify)
    }
  }

  def collect : List[Double] = left.collect ++ right.collect
}

object finalized {
  trait Exp extends exp.m4.Exp[Exp] with Factory {
    def getSelf : Exp = this
  }

  trait BinaryExp extends Exp with exp.m4.BinaryExp[Exp] { }

  trait Factory extends exp.m3.Factory[Exp] {
    override def lit(value: Double) : Exp = new Lit(value)
    override def add(left : exp.Exp[Exp], right : exp.Exp[Exp]) : Exp = new Add(left, right)
    override def sub(left : exp.Exp[Exp], right : exp.Exp[Exp]) : Exp = new Sub(left, right)
    override def mult(left : exp.Exp[Exp], right : exp.Exp[Exp]) : Exp = new Mult(left, right)
    override def neg(inner : exp.Exp[Exp]) : Exp = new Neg(inner)
    override def divd(left : exp.Exp[Exp], right : exp.Exp[Exp]) : Exp = new Divd(left, right)

    implicit override def convert(toConvert : exp.Exp[Exp]) : Exp = toConvert.getSelf
  }

  class Lit(val value : Double) extends Exp with exp.m4.Lit[Exp]
  class Add(var _left : exp.Exp[Exp], var _right : exp.Exp[Exp]) extends exp.m4.Add[Exp] with BinaryExp
  class Sub(var _left : exp.Exp[Exp], var _right : exp.Exp[Exp]) extends exp.m4.Sub[Exp] with BinaryExp
  class Mult(var _left : exp.Exp[Exp], var _right : exp.Exp[Exp]) extends exp.m4.Mult[Exp] with BinaryExp
  class Neg(var _inner : exp.Exp[Exp]) extends exp.m4.Neg[Exp] with Exp
  class Divd(var _left : exp.Exp[Exp], var _right : exp.Exp[Exp]) extends exp.m4.Divd[Exp] with BinaryExp
}

File: coco_csharp/CoCo.Tests/M4.cs
using System;
using Xunit;
using System.Collections.Generic;
using EP.M4;

namespace CoCo.Tests.M4
{
    public interface TestTemplate<FT> : CoCo.Tests.M3.TestTemplate<FT>, Factory<FT> {
        new public void test()
        {
            CoCo.Tests.M3.TestTemplate<FT> parent = this;
            parent.test();

            EP.Exp<FT> expr1 = divd(lit(3), lit(1));
            Assert.Equal("3", convert(convert(expr1).simplify()).prettyp());

            EP.Exp<FT> expr2 = divd(lit(0), lit(2));
            Assert.Equal("0", convert(convert(expr2).simplify()).prettyp());

            EP.Exp<FT> expr3 = divd(lit(3), divd(lit(6.0), lit(2)));
            Assert.Equal("1", convert(convert(expr3).simplify()).prettyp());

            EP.Exp<FT> expr4 = neg(expr1);
            Assert.Equal("(-3)", convert(convert(expr4).simplify()).prettyp());

            EP.Exp<FT> expr6 = add(lit(-2), lit(2));
            Assert.Equal("0", convert(convert(expr6).simplify()).prettyp());

            EP.Exp<FT> expr7 = add(lit(-2), lit(0));
            Assert.Equal("-2", convert(convert(expr7).simplify()).prettyp());

            EP.Exp<FT> expr8 = add(lit(-2), lit(-5));
            Assert.Equal("(-2+-5)", convert(convert(expr8).simplify()).prettyp());

            EP.Exp<FT> sub2 = sub(lit(3), lit(3));
            Assert.Equal("0", convert(convert(sub2).simplify()).prettyp());

            EP.Exp<FT> neg2 = neg(lit(0));
            Assert.Equal("0", convert(convert(neg2).simplify()).prettyp());

            EP.Exp<FT> left = mult(lit(3), lit(1));
            EP.Exp<FT> right = mult(lit(1), lit(9));
            EP.Exp<FT> expr5 = divd(left, right);
            Assert.Equal("(3/9)", convert(convert(expr5).simplify()).prettyp());

            Assert.Equal(convert(divd(divd(lit(5), lit(7.0)), sub(lit(7.0), mult(lit(2), lit(3))))).collect(), new List<double>() {5, 7.0, 7.0, 2, 3});
            Assert.Equal(convert(add(lit(0), lit(0))).collect(), new List<double>() {0, 0});
            Assert.Equal(convert(neg(lit(0))).collect(), new List<double>() { 0 });
            Assert.Equal(convert(mult(lit(1), lit(12))).collect(), new List<double>() {1, 12});
            Assert.Equal(convert(mult(lit(13), lit(1))).collect(), new List<double>() {13, 1});
            Assert.Equal("((5/2)*4)", convert(mult(divd(lit(5), lit(2)), lit(4))).prettyp());
            Assert.Equal(10, convert(mult(divd(lit(5), lit(2)), lit(4))).eval());

            EP.Exp<FT> h1 = mult(lit(3), lit(2));
            EP.Exp<FT> h2 = mult(h1, h1);
            EP.Exp<FT> threeDeep = mult(h2, h2);

            EP.Exp<FT> mult1 = mult(lit(5), threeDeep);
            convert(mult1).truncate(1);
            Assert.Equal("(5*1296)", convert(mult1).prettyp());

            EP.Exp<FT> divd1 = divd(lit(5), threeDeep);
            convert(divd1).truncate(1);
            Assert.Equal("(5/1296)", convert(divd1).prettyp());

            EP.Exp<FT> neg1 = neg(threeDeep);
            convert(neg1).truncate(1);
            Assert.Equal("(-1296)", convert(neg1).prettyp());

            EP.Exp<FT> sub1 = sub(lit(5), threeDeep);
            convert(sub1).truncate(1);
            Assert.Equal("(5-1296)", convert(sub1).prettyp());

            EP.Exp<FT> baseExp = mult(add(lit(3), lit(7.0)), lit(3));
            String baseExpBeforeTrunc = convert(baseExp).prettyp();
            convert(baseExp).truncate(3);
            Assert.Equal(convert(baseExp).prettyp(), baseExpBeforeTrunc);

            EP.Exp<FT> larger = sub(baseExp, divd(baseExp, lit(1)));
            String largerBeforeTrunc = convert(larger).prettyp();
            convert(larger).truncate(4);
            Assert.Equal(convert(larger).prettyp(), largerBeforeTrunc);

            EP.Exp<FT> nd1 = neg(lit(1));
            EP.Exp<FT> nd2 = neg(nd1);
            EP.Exp<FT> nd3 = neg(nd2);
            Assert.Equal("(-(-(-1)))", convert(nd3).prettyp());
            convert(nd3).truncate(2);

            // looks odd, but is right. The innermost one is the negation of -1
            Assert.Equal("(-(--1))", convert(nd3).prettyp());

            convert(larger).truncate(3);
            EP.Exp<FT> baseExpTrunc = mult(lit(10), lit(3));
            Assert.Equal( convert(baseExp).prettyp(), convert(baseExpTrunc).prettyp());
            EP.Exp<FT> largerTrunc = sub(baseExpTrunc, divd(baseExpTrunc, lit(1)));
            Assert.Equal(convert(larger).prettyp(), convert(largerTrunc).prettyp());
        }
    }

    public class ActualTest : TestTemplate<EP.M4.Finalized.Exp>, EP.M4.Finalized.Factory {}

    public class M4Test
    {
        [Fact]
        public void TestM4()
        {
            TestTemplate<EP.M4.Finalized.Exp> theTest = new ActualTest();
            theTest.test();
        }
    }
}
